### ISSUE TITLE
Control parallel processing

### ISSUE THREAD

**Issue by zachokeeffe** [jepusto/clubSandwich#30]
Great package. But I was struggling to find a way, without hacking all of the code, to control how many cores are utilized in simulations. Could you add that, please?

**Comment by jepusto** [comment]
Could you be more specific about what you are trying to do? I'm not sure what functions you are referring to.

**Comment by zachokeeffe** [comment]
With coef_test(lm.fit, vcov="CR2", cluster=data$cluster), for example, it uses all 16 cores on my machine. I'd like to keep a couple free if possible while the models are running.

**Comment by jepusto** [comment]
That's odd. There's nothing in coef_test() or the underlying functions that
is parallelized.

You mentioned simulations---are you running simulations on these functions,
and is this what could be eating all your cores? Or are you getting the
full core utilization even when testing coefficients from one regression
with actual data?

**Comment by zachokeeffe** [comment]
Oh, weird indeed. I was searching this repository for something related to parallelism and saw something about params, so I assumed that must be what it's doing: https://github.com/jepusto/clubSandwich/search?utf8=%E2%9C%93&q=parallel&type=.

No, I'm not doing any simulations. And it isn't running in parallel. This happens when I run a single coef_test on a single regular lm object. Happens in regular R, so it isn't an RStudio thing either. It occurs with the version on CRAN as well as the latest version from GitHub. I'm using Debian 9, R 3.3.3.

**Comment by jepusto** [comment]
None of those search results is part of the actual code base—its all auxiliary files with code for simulations I’ve done to validate the methods (plus there’s one occurrence of “parallel” in a Roxygen comment for a man page).

Could you post a reprex? Does the problem occur if you call vcovCR() directly, or only through coef_test()? Can you isolate it to one of the options for the test = argument?

**Comment by zachokeeffe** [comment]
It occurs if I call vcovCR directly, and the type of test does not matter. If I make a large enough dataset, I see my cores all spike when vcovCR is called.

m <- 1000
cluster <- factor(rep(paste0('a',1:m), 3 + rpois(m, 5)))
n <- length(cluster)
X <- matrix(rnorm(3 * n), n, 3)
nu <- rnorm(m)[cluster]
e <- rnorm(n)
y <- X %*% c(.4, .3, -.3) + nu + e
dat <- data.frame(y, X, cluster, row = 1:n)
nrow(dat)
lm_fit <- lm(y ~ X1 + X2 + X3, data = dat)
vcovCR(lm_fit, cluster = dat$cluster, type = "CR2")

**Comment by jepusto** [comment]
I am not able to reproduce the behavior on a Windows machine.

The only thing that occurs to me is that vcovCR calls Map() for certain
calculations. Do you have R configured somehow to use multiple processes by
default? Is its native behavior different on Debian? Otherwise I'm stumped.

**Comment by zachokeeffe** [comment]
No, I don't have any special set up. Bizarre indeed. I restarted, still does it. And it does it even if parallel isn't loaded, so I'm not even sure how it could be using multiple threads. Now that I'm thinking about it, this could help explain some other weird behavior. For example, I've found it using all my cores to implement glm(). I couldn't find anything online about controlling core usage for that either, probably because it isn't supposed to be using multiple cores. I'm not even sure who the right people to contact about this would be (since I don't know the source of the error), or how I would reproduce it for other people.
